### PR TITLE
Handle navigation from push notifications in Test app

### PR DESCRIPTION
**Pull request by martinmitrevski**
### 🔗 Issue Links

https://stream-io.atlassian.net/browse/CIS-2164

### 🎯 Goal

The goal is to be able to test navigating to specific channels from push notifications in the test app.

Additionally, to bring better alignment between the Test and Demo app, for easier maintaince.

### 📝 Summary

- Support for push notifications tests with deep linking.
- Sharing similar code between the demo and the test app. 

### 🛠 Implementation

- Separated some demo app related code in extensions.
- Common code is shared between demo and test app.
- Differences in the flows are handled in the separate extensions.

### 🧪 Manual Testing Notes

- Test if the demo app is not broken.
- Run the e2e tests to make sure everything is passing.

### ☑️ Contributor Checklist

- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] This change follows zero ⚠️ policy (required)
- [x] This change should be manually QAed
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

Alt Text

### CODE REVIEW

File: StreamChatUITestsAppUITests/Tests/PushNotification_Tests.swift [hunk 1]
```diff
@@ -33,11 +33,9 @@ final class PushNotification_Tests: StreamTestCase {
     
     func test_goToBackgroundFromChannelList_and_tapOnPushNotification() throws {
         linkToScenario(withId: 291)
-        
-        throw XCTSkip("[CIS-2164] The test app is not yet ready for this test")
 
         GIVEN("user goes to channel list") {
-            userRobot.login()
+            userRobot.login().openChannel().tapOnBackButton()
```
Comment (nuno-vieira) [inline StreamChatUITestsAppUITests/Tests/PushNotification_Tests.swift:38]: Why is this needed here? 🤔

File: DemoApp/DemoAppCoordinator+DemoApp.swift [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+// Copyright © 2022 Stream.io Inc. All rights reserved.
+//
+
+import StreamChat
+import StreamChatUI
+import UIKit
+
+// MARK: - Navigation
+
+extension DemoAppCoordinator {
```
Comment (nuno-vieira) [inline DemoApp/DemoAppCoordinator+DemoApp.swift:11]: Maybe it makes sense to create a **DemoAppShared** folder at the root. to be easier to find the code that is shared with the Test App? 🤔

**Review by nuno-vieira — APPROVED** [review]
Overall looks good to me 👍 Just added some minor suggestions, but not blocking at all

File: StreamChatUITestsAppUITests/Tests/PushNotification_Tests.swift (hunk 1 above)
Comment (alteral) [reply StreamChatUITestsAppUITests/Tests/PushNotification_Tests.swift:38]: To make the mock server know which channel the user has entered to send push notifications to that channel.

Comment (nuno-vieira) [reply StreamChatUITestsAppUITests/Tests/PushNotification_Tests.swift:38]: Maybe we can add a comment explaining this? It would be nice

Comment (alteral) [reply StreamChatUITestsAppUITests/Tests/PushNotification_Tests.swift]: Done 👍

File: DemoApp/DemoAppCoordinator+DemoApp.swift (hunk 2 above)
Comment (martinmitrevski) [reply DemoApp/DemoAppCoordinator+DemoApp.swift]: Yes, that makes sense. I've added a Shared folder at the root and organized things a bit better. We can further improve it in the future.
